### PR TITLE
Deprecate libcudf regex APIs accepting pattern strings directly

### PR DESCRIPTION
**Pull request by davidwendt**
## Description
Deprecates the libcudf regex APIs that do not accept `cudf::strings::regex_program` objects.
The libcudf regex functions were converted to accept `regex_program` objects instead of a pattern string and regex-flags directly in PR #11927
Most of this is removing calls to the deprecated functions in the gtests.
The declarations in the headers had to be reordered to make the reference links work in the doxygen output.

These deprecated functions will be removed in a future release.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by davidwendt** [comment]
Apologize for reordering the declarations in the headers to make doxygen work. They headers look difficult to review now. Essentially there are two APIs one with `string_view pattern` parameter and one with `regex_program` parameter. The order of these were swapped in the header file at each location so the `regex_program` one would appear before the `string_view` one.

**Comment by vuule** [comment]
Do you have a timeline to remove the deprecated APIs?
I'm wondering if we should add deprecation warning message(s) once logger is in place, so deprecation is more visible.

**Comment by davidwendt** [comment]
> Do you have a timeline to remove the deprecated APIs? 

I was planning to remove them in 23.06.

> I'm wondering if we should add deprecation warning message(s) once logger is in place, so deprecation is more visible.

The compiler will give a deprecation warning. I'm not sure we should warn actual end users.

**Comment by vuule** [comment]
Oh, right, `[[deprecated]]`. Yeah, that sounds perfect.

**Review by vuule — APPROVED** [review]
doxygen really spiced this one up :D

File: cpp/include/cudf/strings/extract.hpp [hunk 1]
```diff
@@ -27,14 +27,14 @@ namespace strings {
...
- * @addtogroup strings_substring
+ * @addtogroup strings_extract
  * @{
  * @file
  */
 
 /**
  * @brief Returns a table of strings columns where each column corresponds to the matching
- * group specified in the given regular expression pattern.
+ * group specified in the given regex_program object
```
Comment (ttnghia) [inline cpp/include/cudf/strings/extract.hpp:37]:
```suggestion
 * group specified in the given regex_program object.
```

File: cpp/include/cudf/strings/extract.hpp [hunk 2]
```diff
@@ -77,27 +76,31 @@ std::unique_ptr<table> extract(
...
   strings_column_view const& strings,
-  regex_program const& prog,
+  std::string_view pattern,
+  regex_flags const flags             = regex_flags::DEFAULT,
   rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource());
 
 /**
  * @brief Returns a lists column of strings where each string column row corresponds to the
- * matching group specified in the given regular expression pattern.
+ * matching group specified in the given regex_program object
```
Comment (ttnghia) [inline cpp/include/cudf/strings/extract.hpp:103]:
```suggestion
 * matching group specified in the given regex_program object.
```

File: cpp/include/cudf/strings/findall.hpp [hunk 3]
```diff
@@ -33,16 +33,17 @@ struct regex_program;
  */
 
 /**
- * @brief Returns a lists column of strings for each matching occurrence of the
- * regex pattern within each string.
+ * @brief Returns a lists column of strings for each matching occurrence using
+ * the regex_program pattern within each string
```
Comment (ttnghia) [inline cpp/include/cudf/strings/findall.hpp:37]:
```suggestion
 * the regex_program pattern within each string.
```

File: cpp/include/cudf/strings/split/split_re.hpp [hunk 4]
```diff
@@ -34,7 +34,7 @@ struct regex_program;
 
 /**
  * @brief Splits strings elements into a table of strings columns
- * using a regex pattern to delimit each string.
+ * using a regex_program's pattern to delimit each string
```
Comment (ttnghia) [inline cpp/include/cudf/strings/split/split_re.hpp:37]:
```suggestion
 * using a regex_program's pattern to delimit each string.
```

File: cpp/include/cudf/strings/split/split_re.hpp [hunk 5]
```diff
@@ -99,42 +103,41 @@ std::unique_ptr<table> split_re(
...
-  regex_program const& prog,
+  std::string_view pattern,
   size_type maxsplit                  = -1,
   rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource());
 
 /**
- * @brief Splits strings elements into a table of strings columns
- * using a regex pattern to delimit each string starting from the end of the string.
+ * @brief Splits strings elements into a table of strings columns using a
+ * regex_program's pattern to delimit each string starting from the end of the string
```
Comment (ttnghia) [inline cpp/include/cudf/strings/split/split_re.hpp:140]:
```suggestion
 * regex_program's pattern to delimit each string starting from the end of the string.
```

File: cpp/include/cudf/strings/split/split_re.hpp [hunk 6]
```diff
@@ -203,42 +210,41 @@ std::unique_ptr<table> rsplit_re(
...
   strings_column_view const& input,
-  regex_program const& prog,
+  std::string_view pattern,
   size_type maxsplit                  = -1,
   rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource());
 
 /**
  * @brief Splits strings elements into a list column of strings
- * using the given regex pattern to delimit each string.
+ * using the given regex_program to delimit each string
```
Comment (ttnghia) [inline cpp/include/cudf/strings/split/split_re.hpp:247]:
```suggestion
 * using the given regex_program to delimit each string.
```

File: cpp/include/cudf/strings/split/split_re.hpp [hunk 7]
```diff
@@ -307,46 +317,45 @@ std::unique_ptr<column> split_record_re(
...
-  regex_program const& prog,
+  std::string_view pattern,
   size_type maxsplit                  = -1,
   rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource());
 
 /**
- * @brief Splits strings elements into a list column of strings
- * using the given regex pattern to delimit each string starting from the end of the string.
+ * @brief Splits strings elements into a list column of strings using the given
+ * regex_program to delimit each string starting from the end of the string
```
Comment (ttnghia) [inline cpp/include/cudf/strings/split/split_re.hpp:358]:
```suggestion
 * regex_program to delimit each string starting from the end of the string.
```

File: cpp/include/cudf/strings/extract.hpp (hunk 1 above)
Comment (davidwendt) [reply cpp/include/cudf/strings/extract.hpp:37]: These are not sentences and so do not warrant a period.

**Review by ttnghia — APPROVED** [review]

**Pull request merged**